enhancements/cloud-integration/openstack/cloud-provider-config-upgrade.md

```diff
@@ -0,0 +1,432 @@
+- Maintain the interface for configuring the cloud provider, via the
+  `openshift-config/cloud-provider-config` config map.
+
+### Non-Goals
+
+- Modify user-managed configuration
+- Manage cloud providers for other, non-OpenStack platforms
+
+## Proposal
+
+- The change will be implemented in the Cluster Cloud Controller Manager
+  Operator (CCCMO)
+- CCCMO will create a separate config map for the external cloud provider
+- We will prevent upgrade if it is not possible to create the external cloud
+  provider config map
+
+### Implemented in the Cluster Cloud Controller Manager Operator (CCCMO)
+
+The [Cluster Config Operator (CCO)][CCO] is currently responsible for managing
+the config for the legacy cloud provider. This applies a [transformation
+function][transformation-function] to the user-managed config and writes the
+result to `openshift-config-managed/kube-cloud-config`. The transformation
+function is selected according to PlatformType in the cluster Infrastructure
+object. Components requiring legacy cloud provider configuration, including
+kubelet and KCM, read the transformed config map.
+
+CCO will **not** be modified. The new functionality will instead be implemented
+as a change to the [Cluster Cloud Controller Manager Operator (CCCMO)][CCCMO],
+functioning in much the same way as CCO's management of the legacy
+configuration.
+
+[CCO]: https://docs.openshift.com/container-platform/4.9/operators/operator-reference.html#cluster-config-operator_platform-operators-ref
+[transformation-function]: https://github.com/openshift/cluster-config-operator/tree/master/pkg/operator/kube_cloud_config
+[CCCMO]: https://docs.openshift.com/container-platform/4.9/operators/operator-reference.html#cluster-cloud-controller-manager-operator_platform-operators-ref
+
+### Separate ConfigMaps for legacy and external cloud providers
+
+The cloud configuration is read by multiple components which will not all be
+upgraded simultaneously. If we simply updated the
+`openshift-config-managed/kube-cloud-config` to be compatible with external CCM
+we would risk breaking KCM and kubelets which have not yet been upgraded.
+Therefore we are constrained to write two versions of the managed config map.
+
+CCO will be continue to create `openshift-config-managed/kube-cloud-config`,
+which will remain compatible with the legacy cloud provider. CCCMO will create a
+new `openshift-config-managed/cloud-controller-manager-config` config map, which
+will be compatible with the external CCM. The external CCM will then read its
+configuration from `openshift-cloud-controller-manager/cloud-conf`.
+
+In a future release, we can deprecate and remove CCO's management of
+`openshift-config-managed/kube-cloud-config` and eventually the config map
+itself. We can also remove the `openstack-credentials/openstack-credentials`
+secret, which will no longer be used by anything. We will immediately deprecate
+this in the 4.11 release and remove it in the 4.12 release.
```

Comment:
This is unfortunately not that straightforward. We need to keep the in-tree cloud provider running to detach volumes created with the in-tree provisioner.
And because we'll have to support EUS to EUS upgrade (skipping odd release numbers), we'll need to have it in 4.12 as well to detach 4.10 volumes.
Potentially, we can only remove in 4.13.

Reply:
I didn't know/think of that. Updated to reflect this.